src/wallet/mod.rs

```diff
@@ -793,6 +886,93 @@ impl Wallet {
 
 		Ok(tx)
 	}
+
+	fn create_payment_from_tx(
+		&self, locked_wallet: &PersistedWallet<KVStoreWalletPersister>, txid: Txid,
+		payment_id: PaymentId, tx: &Transaction, payment_status: PaymentStatus,
+		confirmation_status: ConfirmationStatus, conflicting_txids: Option<Vec<Txid>>,
+	) -> PaymentDetails {
+		// TODO: It would be great to introduce additional variants for
+		// `ChannelFunding` and `ChannelClosing`. For the former, we could just
+		// take a reference to `ChannelManager` here and check against
+		// `list_channels`. But for the latter the best approach is much less
+		// clear: for force-closes/HTLC spends we should be good querying
+		// `OutputSweeper::tracked_spendable_outputs`, but regular channel closes
+		// (i.e., `SpendableOutputDescriptor::StaticOutput` variants) are directly
+		// spent to a wallet address. The only solution I can come up with is to
+		// create and persist a list of 'static pending outputs' that we could use
+		// here to determine the `PaymentKind`, but that's not really satisfactory, so
+		// we're punting on it until we can come up with a better solution.
+
+		let existing_payment = self.payment_store.get(&payment_id);
+		let final_conflicting_txids = if let Some(provided_conflicts) = conflicting_txids {
+			provided_conflicts
+		} else if let Some(payment) = &existing_payment {
+			if let PaymentKind::Onchain { conflicting_txids: existing_conflicts, .. } =
+				&payment.kind
+			{
+				existing_conflicts.clone()
+			} else {
+				Vec::new()
+			}
+		} else {
+			Vec::new()
+		};
+
+		let kind = crate::payment::PaymentKind::Onchain {
+			txid,
+			status: confirmation_status,
+			conflicting_txids: final_conflicting_txids,
+		};
+
+		let fee = locked_wallet.calculate_fee(tx).unwrap_or(Amount::ZERO);
+		let (sent, received) = locked_wallet.sent_and_received(tx);
+		let fee_sat = fee.to_sat();
+
+		let (direction, amount_msat) = if sent > received {
+			(
+				PaymentDirection::Outbound,
+				Some(
+					(sent.to_sat().saturating_sub(fee_sat).saturating_sub(received.to_sat()))
+						* 1000,
+				),
+			)
+		} else {
+			(
+				PaymentDirection::Inbound,
+				Some(
+					received.to_sat().saturating_sub(sent.to_sat().saturating_sub(fee_sat)) * 1000,
+				),
+			)
+		};
+
+		PaymentDetails::new(
+			payment_id,
+			kind,
+			amount_msat,
+			Some(fee_sat * 1000),
+			direction,
+			payment_status,
+		)
+	}
+
+	fn find_payment_by_txid(&self, target_txid: Txid) -> Option<PaymentId> {
+		let direct_payment_id = PaymentId(target_txid.to_byte_array());
+		if self.payment_store.get(&direct_payment_id).is_some() {
```

Comment:
I think this would unnecessarily clone the entry. Let's just add a `contains` implementation to `DataStore` (preferably in a separate prefactor commit).